### PR TITLE
fix: AppList - namespace filter issues fix

### PR DESCRIPTION
**Pull request by sohel-rp**
# Description
This includes fixes for the following issues -

1. Selecting one more cluster from the list is clearing out all selected namespaces from the filter
2. If multiple clusters are selected & no namespaces are selected for any cluster and applied then it's clearing out the cluster from the filter
3. Removing namespaces from the filter is clearing out the respective cluster when multiple clusters are selected

Fixes - # https://github.com/devtron-labs/devtron/issues/1156

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
All changes have been tested manually.

- By selecting one more cluster when one cluster & multiple namespaces are already been selected -> now it's not clearing out the namespaces from the filter.
-  By selecting multiple clusters & no namespaces for respective clusters and applying them -> now it's not clearing out the clusters from the filter.
- By removing all namespaces one by one from the filter - now it's not clearing out the respective cluster when multiple clusters are selected

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/list-new/AppList.tsx [hunk 1]
```diff
@@ -321,28 +321,67 @@ export default function AppList() {
         history.push(url);
     }
 
-    const applyFilter = (type: string, list: FilterOption[], selectedAppTab : string = undefined): void => {
+    const getAppliedFilters = (ids: (string | number)[], filterType: string, query: Record<string, string>): string => {
```
Comment (manish-agrawal-ai) [inline src/components/app/list-new/AppList.tsx:324]: Please add comment for this function

Comment (sohel-rp) [reply src/components/app/list-new/AppList.tsx:332]: Done

File: src/components/app/list-new/AppList.tsx [hunk 2]
```diff
@@ -321,28 +321,109 @@ export default function AppList() {
...
+        let queryParamType =
+            type == AppListConstants.FilterType.CLUTSER || type == AppListConstants.FilterType.NAMESPACE
+                ? AppListConstants.FilterType.NAMESPACE
+                : type;
+        let checkedItems = list.filter((item) => item.isChecked);
+        let ids = checkedItems.map((item) => item.key);
+
+        query[queryParamType] =
+            queryParamType === AppListConstants.FilterType.NAMESPACE
+                ? getAppliedFilters(ids, type, query)
```
[the remark this answers is not in the document]
Comment (sohel-rp) [reply src/components/app/list-new/AppList.tsx]: Done. Renamed to `getUpdatedFiltersOnNamespaceChange`.

File: src/components/app/list-new/AppList.tsx [hunk 3]
```diff
@@ -321,28 +321,109 @@ export default function AppList() {
...
+    const getUpdatedFiltersOnNamespaceChange = (ids: (string | number)[], filterType: string, query: Record<string, string>): string => {
+        /**
+         * Step 1: Return currently selected/checked items from filters list as string if
+         * - There are no query params
+         * - There is no namespace query param i.e. this is the first time selecting the cluster filter
+         */
+        if (
+            !query ||
+            Object.keys(query).length <= 0 ||
+            typeof query[AppListConstants.FilterType.NAMESPACE] === 'undefined'
```
Comment (manish-agrawal-ai) [inline src/components/app/list-new/AppList.tsx:341]: We should handle empty as well. 
!query[AppListConstants.FilterType.NAMESPACE]

(As after removing the filters, namespace in query becomes empty)

File: src/components/app/list-new/AppList.tsx [hunk 4]
```diff
@@ -321,28 +321,109 @@ export default function AppList() {
...
+            return ids.toString();
+        }
+
+        /**
+         * Step 2: Create & init all required arrays
+         * - appliedFilters: Array of currently applied filters in query param
+         * - checkedItemIds: Array of currently selected/checked items from filters list
+         * - updatedAppliedFilters: Array of new filters to be applied
+         */
+        let appliedFilters = query[AppListConstants.FilterType.NAMESPACE].split(',');
```
Comment (manish-agrawal-ai) [inline src/components/app/list-new/AppList.tsx:352]: currentlyAppliedNamespaceFilters

Comment (sohel-rp) [reply src/components/app/list-new/AppList.tsx]: Done

File: src/components/app/list-new/AppList.tsx (hunk 3 above)
Comment (sohel-rp) [reply src/components/app/list-new/AppList.tsx]: Done

File: src/components/app/list-new/AppList.tsx [hunk 5]
```diff
@@ -328,29 +328,110 @@ export default function AppList() {
...
+            _currentTab == AppListConstants.AppTabs.DEVTRON_APPS ? buildDevtronAppListUrl() : buildHelmAppListUrl()
+        }?${queryStr}`;
         history.push(url);
-    }
+    };
 
     const removeFilter = (filter, filterType: string): void => {
-        let val = filter.key;
+        let val = filter.key.toString();
+        let clustId = val.split('_')[0];
```
Comment (manish-agrawal-ai) [inline src/components/app/list-new/AppList.tsx:434]: move loacally

Comment (sohel-rp) [reply src/components/app/list-new/AppList.tsx]: Can't move it locally as it's used at 2 places so kept it at the top & added a comment as suggested.

**Review by manish-agrawal-ai — APPROVED** [review]
approved

**Pull request merged**